### ISSUE TITLE
When host silently disconnect then the game is not recoverable (GNA #24042)

### ISSUE THREAD

**Issue by wesnoth-bugs** [wesnoth/wesnoth#1521]
**[Original submission](https://gna.org/bugs/index.php?24042)** by neverenough on 2015-11-07

See also #24041

When game host silently disconnect there is no way to force other clients to become host (neither rejoining user will be able to have controll back).

The "ping" command suggested in #24041 could be a timed trigger to move game's control from previous host to next player (as in normal leaves). 

Unlike in #24041 the implementation _must_ involve the server as third party check to avoid abuses

(Reproduced on any)
Release: any
Priority: 5 - Normal
Severity: 3 - Normal

**Comment by wesnoth-bugs** [comment]
**Modified on 2015-11-09**

**soliton wrote:**

> What is a "silent disconnect"?
> 
> 
> 
> If you mean going afk or similar then there is no technical solution to that. Save the game and reload with the original host replaced.

**Comment by wesnoth-bugs** [comment]
**Modified on 2015-11-10**

**neverenough wrote:**

> This happened to me: host was apparently afk for long time, then he rejoined with "nick1" (unregistered player) and told he had a disconnection.
> 
> I guess probably he had some network related issue, i.e. ISP disconnection, wlan trouble, or maybe battery laptop discharge.
> 
> I can't try to reproduce alone, would need a remote tester. Anyway looks reasonable that something is wrong with tcp ping timeout network setting.
> 
> 
> 
> I'm not sure what would happen if host was using a _registered_ nick, i guess probably server would kick the old nick when he tried to rejoin and switch host controls to second player in game (worth a check anyway).
> 
> 
> 
> This behaviour can't unluckily get executed for unregistered people.

**Comment by wesnoth-bugs** [comment]
**Modified on 2015-12-04**

**gfgtdf wrote:**

> I also was in this situation (had a dc i think it was the daily ip reset recomment at 3:00 but not 100%) sure and when i tried to logged in i noticed that my nick was still treated as logged in by the server.
> 
> 
> 
> What i suggest is that when one logs in with a registered nick which is already online that the other connection gets kicked and the player can log in.

**Comment by wesnoth-bugs** [comment]
**Modified on 2015-12-04**

**gfgtdf wrote:**

> unfortulateley this wouldn't solve the issue for unregistered nicks though.

**Comment by wesnoth-bugs** [comment]
**Modified on 2015-12-28**

**coffee wrote:**

> I am happy to assist with testing this bug, as I can attest that it is actually quite common and has a severe impact on games.
> 
> 
> 
> Although probably not that useful, I've done some tests with the linux "tc" command to simulate packet loss, reordered/corrupt packets, lag, and combinations of the above on a locally hosted mp server with 2 different network interface cards. I couldn't reproduce this effect.
> 
> 
> 
> One thing I wonder, because of my own experience with this bug, is if it happens when only 2 players are playing or does this need 3 or more players and how can this kind of bug be tested easily?

**Comment by wesnoth-bugs** [comment]
**Modified on 2015-12-28**

**coffee wrote:**

> Apologies, I may have put a red herring in there with the number of players. This appears to not to have any bearing on this bug as I have seen it with all different numbers of players.

**Comment by wesnoth-bugs** [comment]
**Modified on 2015-12-29**

**coffee wrote:**

> gfgtdf might be on to something with the ip change. I can reproduce what appears to be the exact same conditions by changing the IP address mid-game of one of the clients.
> 
> 
> 
> When quitting the game it does the same thing as on the mp online server each time I've experienced this (showing out of date information for the semi-d/c player), etc. This makes sense as ISPs change the IP addresses every so often for most people.
> 
> 
> 
> The next thing I figure would be to work out how to detect this in-game.

**Comment by wesnoth-bugs** [comment]
**Modified on 2015-12-30**

**neverenough wrote:**

> ok i marked this a feature request, but looks more like a bug.
> 
> 
> 
> The server doesn't notice silent disconnections.
> 
> 
> 
> coffee u are right: whenever i see this behaviour there is no server message "nick1 has same ip as nick", people got their own connectivity problem and rejoin the game with new ip and nick.
> 
> 
> 
> I've found in server code a section about "ghost players", which *i think* gets filled with clients that don't have "ping timeout" value set. This is default setting for any client.
> 
> Houston, we have a problem

**Comment by wesnoth-bugs** [comment]
**Modified on 2016-01-02**

**coffee wrote:**

> I've done a bit more research into this bug and found that by setting the "ping timeout" value in the preferences the client can detect if it is no longer connected after the specified time. However, the other players who are still connected to the original game will not detect that this player has disconnected.
> 
> 
> 
> In the source code file network_worker.cpp there is code to deal with disconnects in "receive_with_timeout" functions, etc. This code runs if the original client that disconnects connection re-establishes even if they have quit the game.
> 
> 
> 
> Hence, you might think that in the "receive_bytes" function SDLNet_TCP_Recv might be a blocking call, waiting like a zombie to get information on a separate thread. This doesn't appear to be the case, because I have refactored my local version to do a SDLNet_CheckSockets and SDLNet_SocketReady and there is no problem here with blocking that I can see as this code is not run when there is a disconnect and only when the connection is re-established does it run again.
> 
> 
> 
> This is highly frustrating to debug, but perhaps it could be solved by firstly making the clients use the ping test "timeout" to the server as in the current code for the client side. Then a test to each client from the server might be needed as well to disconnect forcefully any client that isn't responding any more.
> 
> 
> 
> When a client disconnects, perhaps they should be given an option to reconnect with the same username and have the server do a check for any open connections to that username, and then a second check  after a period of time to make sure they are disconnected. Once this has been confirmed, then disconnect all sockets with the old connection and allow login with the same handle.

**Comment by wesnoth-bugs** [comment]
**Modified on 2016-01-02**

**gfgtdf wrote:**

> We are currently planning to move the  netwokring code from SDL to boost::asio, so i tink we shouldnt change that part of thew code now.

**Comment by wesnoth-bugs** [comment]
**Modified on 2016-01-02**

**coffee wrote:**

> Good to know, and happy to help test this again when the code gets replaced.

**Comment by Wedge009** [comment]
@gfgtdf, @Coffee--, @soliton- We have moved network code to boost::asio some time ago, didn't we? Is it time to re-test this?

**Comment by gfgtdf** [comment]
i wouldnt know how, @loonycyborg do you remember whether this was fixed?

**Comment by loonycyborg** [comment]
If I understand this issue correctly then it's rendered moot by mandatory registration on the official server. If host that had a disconnect relogs he'll just autokick his previous disconnected iteration. Though if host gets a dc and goes missing it will be harder to retake control of the game if network layer won't detect the dc, since keep-alive timeout is very long. In that case another player can save and load the game with himself as host.

**Comment by Wedge009** [comment]
Then perhaps that just leaves the unregistered host case on non-official servers?

**Comment by AI0867** [comment]
Is the 'host' here also the machine with wesnothd? Because in that case, the "guest client" tries to contact wesnothd, fails to reach it and disconnects from *wesnothd*. This issue is about the "host client" disconnecting from wesnothd and wesnothd not realizing this, and therefore not notifying the other clients.

**Comment by soliton-** [comment]
The keep-alive feature is to detect disconnects without the need to explicitely send data. However even if it's the disconnected player's turn you could send a chat message in this case. There might be other cases where you cannot trigger any data sending though I can't think of one from the top of my head.

I think this can be closed unless there is evidence of a true undetected disconnect.

**Comment by Wedge009** [comment]
Originally I did test with the game host on the same PC as the wesnothd instance.

I also just tried wesnothd being on a _different_ PC to the game host: the game host can recover since it has the option to save the game and has an error message 'The semaphore timeout period has expired'. The guest, however, remains in limbo for about 20 minutes before hitting the time-out, which unfortunately is quite a bit longer than the first case where the game host was also running wesnothd. Nonetheless, the disconnection is detected eventually so it's not a true undetected disconnection so closing this.

(Maybe a shorter time-out limit might be preferable but I think that's outside the scope of this issue.)
